Sema: implement `getInitKind` using a request evaluator.

hborla opened the pull request:
* Parse: check if constructor decls are marked as  `convenience` rather than (indirectly) calling `getInitKind`
  * We don't want to invoke the `InitKindRequest` request during parsing. What the parsing code really cares about is whether or not the decl has `ConvenienceAttr`.
* Sema: don't change the init kind of a `ConstructorDecl` as a part of error recovery when type checking its body.
* Sema: simplify checking for a designated init in `canSuppressPotentialWitnessWarningWithMovement`
* Add the request `InitKindRequest` to lazily compute the kind of an initializer declaration.

hborla committed:
Parse: check if constructor decls are marked as  `convenience` rather than (indirectly) calling `getInitKind`

When we requestify `getInitKind`, we don't want to invoke the request during parsing. What the parsing code
really cares about is whether or not the decl has `ConvenienceAttr`.

hborla committed:
Sema: don't change the init kind of a `ConstructorDecl` when type checking its body.

hborla committed:
Sema: simplify checking for a designated init in `canSuppressPotentialWitnessWarningWithMovement`

hborla committed:
Sema: implement `getInitKind` using a request evaluator.

Add the request `InitKind` to lazily compute the kind of an initializer
declaration.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

DougGregor commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -1228,6 +1228,65 @@ static bool doesAccessorNeedDynamicAttribute(AccessorDecl *accessor) {
+
+  // Convenience inits are only allowed on classes and in extensions thereof.
+  if (decl->getAttrs().hasAttribute<ConvenienceAttr>()) {
+    if (auto extType = decl->getDeclContext()->getDeclaredInterfaceType()) {
```
I suggest using

```
if (auto nominal = decl->getDeclContext()->getSelfNominalTypeDecl()) {
```

here instead of `getDeclaredInterfaceType()`, because it's cheaper to evaluate (name lookup vs. type checking). You can then `dyn_cast<ClassDecl>(nominal)` to get `extClass`, and most of the rest of the code will stay the same.

DougGregor reviewed (approved):
This looks great!

hborla committed:
Sema: small stylistic change in `InitKindRequest::evaluate`

Don't use `else` after a `return`.

hborla committed:
Sema: compute `InitKindRequest::evaluate` in a more performant way.

Use `getSelfNominalTypeDecl` rather than `getDeclaredInterfaceType` in
`InitKindRequest::evaluate`, since it's cheaper to perform.

hborla committed:
Merge remote-tracking branch 'upstream/master' into init-kind-request

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility
